File: sol/solution/0101-0200/0109/Solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func sortedListToBST(_ head: ListNode?) -> TreeNode? {
        if head == nil { return nil }
        if head?.next == nil {
            return TreeNode(head!.val)
        }
        var slow = head, fast = head, prev: ListNode? = nil
        while fast != nil && fast?.next != nil {
            prev = slow
            slow = slow?.next
            fast = fast?.next?.next
        }
        prev?.next = nil
        let root = TreeNode(slow!.val)
        root.left = sortedListToBST(head === slow ? nil : head)
        root.right = sortedListToBST(slow?.next)
        return root
    }
}

File: sol/solution/0101-0200/0109/Solution.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  TreeNode? sortedListToBST(ListNode? head) {
    if (head == null) return null;
    if (head.next == null) return TreeNode(head.val);
    ListNode? slow = head, fast = head, prev;
    while (fast?.next != null && fast?.next?.next != null) {
      prev = slow;
      slow = slow?.next;
      fast = fast?.next?.next;
    }
    prev?.next = null;
    TreeNode root = TreeNode(slow!.val);
    root.left = sortedListToBST(head == slow ? null : head);
    root.right = sortedListToBST(slow.next);
    return root;
  }
}